drivers/can/can_stm32.c

```diff
@@ -1163,132 +1151,81 @@ static const struct can_driver_api can_api_funcs = {
+{                                                                    \
+	IRQ_CONNECT(DT_INST_IRQ_BY_NAME(inst, rx0, irq),             \
+		    DT_INST_IRQ_BY_NAME(inst, rx0, priority),        \
+		    can_stm32_rx_isr, DEVICE_DT_INST_GET(inst), 0);  \
+	irq_enable(DT_INST_IRQ_BY_NAME(inst, rx0, irq));             \
+	IRQ_CONNECT(DT_INST_IRQ_BY_NAME(inst, tx, irq),              \
+		    DT_INST_IRQ_BY_NAME(inst, tx, priority),         \
+		    can_stm32_tx_isr, DEVICE_DT_INST_GET(inst), 0);  \
+	irq_enable(DT_INST_IRQ_BY_NAME(inst, tx, irq));              \
+	IRQ_CONNECT(DT_INST_IRQ_BY_NAME(inst, sce, irq),             \
+		    DT_INST_IRQ_BY_NAME(inst, sce, priority),        \
+		    can_stm32_state_change_isr,                      \
+		    DEVICE_DT_INST_GET(inst), 0);                    \
+	irq_enable(DT_INST_IRQ_BY_NAME(inst, sce, irq));             \
+	can->IER |= CAN_IER_TMEIE | CAN_IER_ERRIE | CAN_IER_FMPIE0 | \
+		    CAN_IER_FMPIE1 | CAN_IER_BOFIE;                  \
+	if (IS_ENABLED(CONFIG_CAN_STATS)) {                          \
+		can->IER |= CAN_IER_LECIE;                           \
+	}                                                            \
 }
-
-#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(can1), okay) */
-
-#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(can2), st_stm32_can, okay)
-
-static void config_can_2_irq(CAN_TypeDef *can);
-
-PINCTRL_DT_DEFINE(DT_NODELABEL(can2));
-
-static const struct can_stm32_config can_stm32_cfg_2 = {
-	.can = (CAN_TypeDef *)DT_REG_ADDR(DT_NODELABEL(can2)),
-	.master_can = (CAN_TypeDef *)DT_PROP(DT_NODELABEL(can2),
-								master_can_reg),
-	.bus_speed = DT_PROP(DT_NODELABEL(can2), bus_speed),
-	.sample_point = DT_PROP_OR(DT_NODELABEL(can2), sample_point, 0),
-	.sjw = DT_PROP_OR(DT_NODELABEL(can2), sjw, 1),
-	.prop_ts1 = DT_PROP_OR(DT_NODELABEL(can2), prop_seg, 0) +
-		    DT_PROP_OR(DT_NODELABEL(can2), phase_seg1, 0),
-	.ts2 = DT_PROP_OR(DT_NODELABEL(can2), phase_seg2, 0),
-	.one_shot = DT_PROP(DT_NODELABEL(can2), one_shot),
-	.pclken = {
-		/* can1 (master) clock must be enabled for can2 as well */
-		.enr = DT_CLOCKS_CELL(DT_NODELABEL(can1), bits) |
-		       DT_CLOCKS_CELL(DT_NODELABEL(can2), bits),
-		.bus = DT_CLOCKS_CELL(DT_NODELABEL(can2), bus),
-	},
-	.config_irq = config_can_2_irq,
-	.pcfg = PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(can2)),
-	.phy = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(DT_NODELABEL(can2), phys)),
-	.max_bitrate = DT_CAN_TRANSCEIVER_MAX_BITRATE(DT_NODELABEL(can2), 1000000),
+#endif /* CONFIG_SOC_SERIES_STM32F0X */
+
+#define CAN_STM32_CONFIG_INST(inst)                                      \
+PINCTRL_DT_INST_DEFINE(inst);                                            \
+static const struct can_stm32_config can_stm32_cfg_##inst = {            \
+	.can = (CAN_TypeDef *)DT_INST_REG_ADDR(inst),                    \
+	.master_can = (CAN_TypeDef *)DT_INST_PROP_OR(inst,               \
+		master_can_reg, DT_INST_REG_ADDR(inst)),                 \
```

Comment:
Wouldn't this register _always_ be the reg base of the `can1` DTS node? Can we get rid of the separate DTS property for it now?

Reply:
No, we should keep it, as CAN3 is an independent peripheral again and should work as well with this approach. Just couldn't test because I dont have a board with 3 CANs.